admission-webhook/integration_tests/integration_test.go

```diff
@@ -0,0 +1,331 @@
+
+	testConfig, tearDownFunc := integrationTestSetup(t, testName, credSpecTemplates, templates)
+	defer tearDownFunc()
+
+	pod := waitForPodToComeUp(t, testConfig.Namespace, "app="+testName)
+
+	assert.Equal(t, expectedCredSpec0, pod.Annotations["nginx0.container.alpha.windows.kubernetes.io/gmsa-credential-spec"])
+	assert.Equal(t, expectedCredSpec1, pod.Annotations["pod.alpha.windows.kubernetes.io/gmsa-credential-spec"])
+	assert.Equal(t, expectedCredSpec2, pod.Annotations["nginx2.container.alpha.windows.kubernetes.io/gmsa-credential-spec"])
+}
+
+func TestHappyPathWithPreSetMatchingAnnotations(t *testing.T) {
+	testName := "happy-path-with-pre-set-matching-annotations"
+	credSpecTemplates := []string{"credspec-0"}
+	templates := []string{"credspecs-users-rbac-role", "service-account", "sa-rbac-binding", "simple-with-gmsa"}
+
+	testConfig, tearDownFunc := integrationTestSetup(t, testName, credSpecTemplates, templates)
+	defer tearDownFunc()
+
+	pod := waitForPodToComeUp(t, testConfig.Namespace, "app="+testName)
+
+	assert.Equal(t, expectedCredSpec0, pod.Annotations["pod.alpha.windows.kubernetes.io/gmsa-credential-spec"])
+}
+
+func TestServiceAccountDoesNotHavePermissionsToUseCredSpec(t *testing.T) {
+	testName := "sa-does-not-have-permissions-to-use-cred-spec"
+	credSpecTemplates := []string{"credspec-0"}
+	templates := []string{"credspecs-users-rbac-role", "service-account", "simple-with-container-level-gmsa"}
+
+	testConfig, tearDownFunc := integrationTestSetup(t, testName, credSpecTemplates, templates)
+	defer tearDownFunc()
+
+	replicaSet := waitForReplicaSetGen1(t, testConfig.Namespace, "app="+testName)
+	assert.Equal(t, int32(0), replicaSet.Status.Replicas)
+	if assert.Equal(t, 1, len(replicaSet.Status.Conditions)) {
+		condition := replicaSet.Status.Conditions[0]
+
+		assert.Equal(t, condition.Reason, "FailedCreate")
+
+		expectedSubstr := fmt.Sprintf("service account %s is not authorized `use` gMSA cred spec %s", testConfig.ServiceAccountName, testConfig.CredSpecNames[0])
+		assert.Contains(t, condition.Message, expectedSubstr)
+	}
+}
+
+func TestCredSpecDoesNotExist(t *testing.T) {
+	testName := "cred-spec-does-not-exist"
+	templates := []string{"all-credspecs-users-rbac-role", "service-account", "sa-rbac-binding", "simple-with-unknown-gmsa"}
+
+	testConfig, tearDownFunc := integrationTestSetup(t, testName, nil, templates)
+	defer tearDownFunc()
+
+	replicaSet := waitForReplicaSetGen1(t, testConfig.Namespace, "app="+testName)
+	assert.Equal(t, int32(0), replicaSet.Status.Replicas)
+	if assert.Equal(t, 1, len(replicaSet.Status.Conditions)) {
+		condition := replicaSet.Status.Conditions[0]
+
+		assert.Equal(t, condition.Reason, "FailedCreate")
+
+		assert.Contains(t, condition.Message, "cred spec i-sure-dont-exist does not exist")
+	}
+}
+
+func TestCannotPreSetGMSAPodLevelContentsAnnotationsWithoutNameAnnotations(t *testing.T) {
```

Comment:
Not sure if this is covered but we should have a negative case where Content Annotations are set and  corresponding Name Annotations are also set but the actual contents of the credspec pointed to by Name Annotations differ from what is specified in Content Annotations.

Reply:
Good call, adding.